okteto/okteto
# Deploy watch refactor

# Proposed changes

This PR is open to discuss the possible refactor of the watch mechanism on the deploy cmd
Following up this PR https://github.com/okteto/okteto/pull/4547

Additionally, i spotted a wierd behavour when certain timeouts where use, i added two additional commits to discuss a possible improvement.

Currently, the wait implementation is done with a ticker every 5 seconds that does a polling of deployments and statefulsets, looking if their replicas are running and skipping when all al running.

Whenever the timeout is set lower than 5 seconds, the process exits with error, but there havent been a call to check the state of the deployment. 

- OPTION1 To solve this my proposal on this commit https://github.com/okteto/okteto/commit/90f22f0c6daf6d04faf6730b4ea6099967a48b77 is to use context with Timeout and wait after the first iteration if there are resources not running, maintaining the 5 second polling. Once the timeout reaches the deadline its canceled and if not finished err is returned.

Also, the mechanism is calling twice to k8s api, one for retrieving deployments and statfulsets list, other to retrieve the deployment or statefulset.

- OPTION 2 watch the deployments and sfs. With this approach, the context used on the k8s pod watch has a timeout deadline so it returns err if not finish before that. While the pods change from its initial state (resourceVersion) the cli will listen for all to be running.


## How to validate

- Compile the binary and run deployments with timeouts/default timeouts
- Check the status of the deployment is correctly updated when running okteto deploy

## CLI Quality Reminders 🔧

For both authors and reviewers:

- Scrutinize for potential regressions
- Ensure key automated tests are in place
- Build the CLI and test using the validation steps
- Assess Developer Experience impact (log messages, performances, etc)
- If too broad, consider breaking into smaller PRs
- Adhere to our [code style](https://github.com/okteto/okteto/blob/master/docs/code-style.md) and [code review](https://github.com/okteto/okteto/blob/master/docs/code-review.md) guidelines
